web/web.go

```diff
@@ -838,3 +842,82 @@ func getAccessToken(c *api.Context, w http.ResponseWriter, r *http.Request) {
 
 	w.Write([]byte(accessRsp.ToJson()))
 }
+
+func incomingWebhook(c *api.Context, w http.ResponseWriter, r *http.Request) {
+	params := mux.Vars(r)
+	id := params["id"]
+
+	hchan := api.Srv.Store.Webhook().GetIncoming(id)
+
+	r.ParseForm()
+
+	props := model.MapFromJson(strings.NewReader(r.FormValue("payload")))
+
+	text := props["text"]
+	if len(text) == 0 {
+		http.Error(w, "No text specified", 500)
+	}
+
+	channelName := props["channel"]
+
+	var hook *model.IncomingWebhook
+	if result := <-hchan; result.Err != nil {
+		l4g.Error(result.Err)
+		http.Error(w, "Invalid webhook", 500)
+		return
+	} else {
+		hook = result.Data.(*model.IncomingWebhook)
+	}
+
+	var channel *model.Channel
+	var cchan store.StoreChannel
+
+	if len(channelName) != 0 {
+		if channelName[0] == '@' {
+			if result := <-api.Srv.Store.User().GetByUsername(hook.TeamId, channelName[1:]); result.Err != nil {
+				l4g.Error(result.Err)
+				http.Error(w, "Couldn't find the user", 500)
+				return
+			} else {
+				channelName = model.GetDMNameFromIds(result.Data.(*model.User).Id, hook.UserId)
+			}
+		} else if channelName[0] == '#' {
+			channelName = channelName[1:]
+		}
+
+		cchan = api.Srv.Store.Channel().GetByName(hook.TeamId, channelName)
+	} else {
+		cchan = api.Srv.Store.Channel().Get(hook.ChannelId)
+	}
+
+	// parse links into Markdown format
+	linkWithTextRegex := regexp.MustCompile(`<([^<\|]+)\|([^>]+)>`)
+	text = linkWithTextRegex.ReplaceAllString(text, "[${2}](${1})")
+
+	linkRegex := regexp.MustCompile(`<\s*(\S*)\s*>`)
+	text = linkRegex.ReplaceAllString(text, "${1}")
+
+	if result := <-cchan; result.Err != nil {
+		l4g.Error(result.Err)
+		http.Error(w, "Couldn't find the channel", 500)
+		return
+	} else {
+		channel = result.Data.(*model.Channel)
+	}
+
+	pchan := api.Srv.Store.Channel().CheckPermissionsTo(hook.TeamId, channel.Id, hook.UserId)
+
+	post := &model.Post{UserId: hook.UserId, ChannelId: channel.Id, Message: text}
+
+	if !c.HasPermissionsToChannel(pchan, "createIncomingHook") && channel.Type != model.CHANNEL_OPEN {
```

Comment:
The first condition checks if the user is part of the channel. If the user is not part of the channel, but the channel is open, then we still want the post to go through (hence the second condition).